internal/build/imgsrc/ensure_builder.go

```diff
@@ -0,0 +1,325 @@
+			span.AddEvent(fmt.Sprintf("non-server error %d", flapsErr.ResponseStatusCode))
+			numRetries += 1
+
+			if numRetries >= 3 {
+				tracing.RecordError(span, err, "error listing machines")
+				return nil, err
+			}
+			time.Sleep(1 * time.Second)
+		} else {
+			tracing.RecordError(span, err, "error listing machines")
+			return nil, err
+		}
+	}
+
+	if len(machines) != 1 {
+		span.AddEvent(fmt.Sprintf("invalid machine count %d", len(machines)))
+		tracing.RecordError(span, InvalidMachineCount, "the existing builder app has an invalid number of machines")
+		return nil, InvalidMachineCount
+	}
+
+	return machines[0], nil
+}
+
+func createBuilder(ctx context.Context, org *fly.Organization, region, builderName string) (app *fly.App, mach *fly.Machine, err error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "create_builder")
+	defer span.End()
+
+	client := flyutil.ClientFromContext(ctx)
+	flapsClient := flapsutil.ClientFromContext(ctx)
+
+	app, err = client.CreateApp(ctx, fly.CreateAppInput{
+		OrganizationID:  org.ID,
+		Name:            builderName,
+		AppRoleID:       "remote-docker-builder",
+		Machines:        true,
+		PreferredRegion: fly.StringPointer(region),
+	})
+	if err != nil {
+		tracing.RecordError(span, err, "error creating app")
+		return nil, nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			span.AddEvent("cleaning up new builder app due to error")
+			client.DeleteApp(ctx, builderName)
+		}
+	}()
+
+	_, err = client.AllocateIPAddress(ctx, app.Name, "shared_v4", "", org, "")
+	if err != nil {
+		tracing.RecordError(span, err, "error allocating ip address")
+		return nil, nil, err
+	}
+
+	guest := fly.MachineGuest{
+		CPUKind:  "shared",
+		CPUs:     4,
+		MemoryMB: 4096,
+	}
+
+	err = flapsClient.WaitForApp(ctx, app.Name)
+	if err != nil {
+		tracing.RecordError(span, err, "error waiting for builder")
+		return nil, nil, fmt.Errorf("waiting for app %s: %w", app.Name, err)
+	}
+
+	var volume *fly.Volume
+	numRetries := 0
+	for {
+		volume, err = flapsClient.CreateVolume(ctx, fly.CreateVolumeRequest{
+			Name:                "machine_data",
+			SizeGb:              fly.IntPointer(50),
+			AutoBackupEnabled:   fly.BoolPointer(false),
+			ComputeRequirements: &guest,
+			Region:              region,
+		})
+		if err == nil {
+			break
+		}
+
+		var flapsErr *flaps.FlapsError
+		if errors.As(err, &flapsErr) && flapsErr.ResponseStatusCode >= 500 && flapsErr.ResponseStatusCode < 600 {
+			span.AddEvent(fmt.Sprintf("non-server error %d", flapsErr.ResponseStatusCode))
+			numRetries += 1
+
+			if numRetries >= 5 {
+				tracing.RecordError(span, err, "error creating volume")
+				return nil, nil, err
+			}
+			time.Sleep(1 * time.Second)
+		} else {
+			tracing.RecordError(span, err, "error creating volume")
+			return nil, nil, err
+		}
+	}
+
+	defer func() {
+		if err != nil {
+			span.AddEvent("cleaning up new volume due to error")
+			flapsClient.DeleteVolume(ctx, volume.ID)
+		}
```

Comment:
It might be better to name the return error something like `retErr` so we don't accidentally introduce a bug by shadowing the `err` elsewhere in the future.

Reply:
done :)